Fram backend

jauzatopauza opened the pull request:
A Fram backend that fits the new architecture. 

I have added the directories `test/sexpr` and `test/calc` with some Fram tests. The duration of effect reconstruction in the latter remains very long.

Another apparent issue is that the generated parsers often contain redundant pattern-matching cases. The OCaml parsers handle this by muting the warnings in their code. Have we got such a feature yet?

jauzatopauza committed:
Fram backend

- Enabled Printer.ml to output Fram syntax.
- Added a Fram code generation module.
- Re-added tests from old Fram backend. Use test/sexpr to witness
  that something works. Effect reconstruction in test/calc still takes
  too long.
- Re-added framtools with the Parsing module.

ppolesiuk commented:
I tried to play with the Fram backend a bit. With the current implementation of effect inference, it requires ~10 minutes to type-check the generated parser of Lua, but I prepared an optimization [#294](https://github.com/fram-lang/dbl/pull/294) that shortens this time to around 5 seconds.

I noticed one important issue with this backend that should be fixed: there is no way of importing a module from the generated code. The `%{ import M %}` doesn't work, since this code is appended after some generated preamble, that contains some definitions, but Fram requires that all imports must appear before definitions.

ppolesiuk commented on framtools/Parsing.fram:
```diff
@@ -0,0 +1,24 @@
+{# This file should be placed in the same directory as the generated
+   parser #}
```
It is not necessary. This file could be a library, that is accessible through for the Fram compiler/interpreter. DBL supports `-L` flag.

ppolesiuk commented on framtools/Parsing.fram:
```diff
@@ -0,0 +1,24 @@
+    { fname : String
+    , lnum : Int
+    , bol : Int
+    , cnum : Int }
```
1. Fram supports record type. Mentioning the constructor name explicitly is not necessary.
2. Code formatting could be improved (closing brace in a new line, 2 space indentation) to match existing code base, as well as proposed coding style.

```suggestion
pub data Pos =
  { fname : String
  , lnum : Int
  , bol : Int
  , cnum : Int
  }
```

ppolesiuk commented on framtools/Parsing.fram:
```diff
@@ -0,0 +1,24 @@
+pub data Lex E Tok = 
+    { token  : Unit ->[E] Tok
+    , startPos : Unit ->[E] Pos
+    , curPos : Unit -> [E] Pos }
```
Same as above.
```suggestion
pub data Lex E Tok = 
  { token  : Unit ->[E] Tok
  , startPos : Unit ->[E] Pos
  , curPos : Unit -> [E] Pos
  }
```

ppolesiuk commented on lib/Backend/Fram.ml:
```diff
@@ -0,0 +1,454 @@
+import Parsing
+import List
+parameter E_err
+parameter ~error : Parsing.Error E_err
```
The declaration of parameters should be moved somewhere below. The prelude should include only imports, so the user can provide own imports before the first definition (Fram requires that all imports are before definitions).

ppolesiuk commented on lib/Backend/Fram.ml:
```diff
@@ -0,0 +1,454 @@
+match ~loc with
+| l :: _ => snd l
+| [] => Parsing.dummyPos
+end
```
Shouldn't we indent this code?

```suggestion
  match ~loc with
  | l :: _ => snd l
  | [] => Parsing.dummyPos
  end
```

ppolesiuk commented on lib/Backend/Fram.ml:
```diff
@@ -0,0 +1,454 @@
+match List.nth ~loc (n - 1) with
+| Some l => fst l
+| None => _kw_endpos n
+end
```
As above.

```suggestion
  match List.nth ~loc (n - 1) with
  | Some l => fst l
  | None => _kw_endpos n
  end
```

ppolesiuk commented on lib/Backend/Fram.ml:
```diff
@@ -0,0 +1,454 @@
+let state_lib = {|
+let lexfun {E_err, E_lex,
+            ~error : Parsing.Error E_err,
+            ~lex : Parsing.Lex E_lex Tok} ppos = 
```
I would format this code like the following.
```suggestion
let lexfun
    {E_err, E_lex
    , ~error : Parsing.Error E_err
    , ~lex : Parsing.Lex E_lex Tok
    } ppos = 
```

ppolesiuk commented on lib/Backend/Fram.ml:
```diff
@@ -0,0 +1,454 @@
+
+pub let shift {E_err, E_lex,
+           ~error : Parsing.Error E_err,
+           ~lex : Parsing.Lex E_lex Tok} () = 
```
Same as above.

ppolesiuk commented on lib/Backend/Fram.ml:
```diff
@@ -0,0 +1,454 @@
+
+let locReduce {E_err, E_lex,
+                ~error : Parsing.Error E_err,
+                ~lex : Parsing.Lex E_lex Tok} n =
```
Also, code formatting.

ppolesiuk commented on lib/Backend/Printer.ml:
```diff
@@ -238,3 +238,220 @@ module Ml (S : Types.BackEndSettings) = struct
+    output_string " = ";
+    if bl then pp_block_expr expr else pp_expr expr
+  
+  (* We have no disjunctions of patterns yet.
+     For now, this function copies the same right-hand-side expression
+     for each left-hand-side pattern in `patterns`. *)
```
We have no or patterns yet, but hopefully, we would have them in the near future. See [#287](https://github.com/fram-lang/dbl/pull/287).

ppolesiuk commented on test/calc/Calc.fram:
```diff
@@ -0,0 +1,41 @@
+	    , startPos = effect () / r => fn ys => r Parsing.dummyPos ys }
+	    return x => fn _ => x
+	    finally f => f xs
+	in f lex
```
In Fram, we use two-space indentation.

ppolesiuk commented on test/calc/Parser.mly:
```diff
@@ -0,0 +1,36 @@
+	if n == 0 then 1
+	else if n == 1 then a
+	else (let (b : Int) = pow a (n / 2) in
+	b * b * (if n % 2 == 0 then 1 else a))
```
Fix indentation.

ppolesiuk commented on test/sexpr/SExpr.fram:
```diff
@@ -0,0 +1,57 @@
+	    , startPos = effect () / r => fn ys => r Parsing.dummyPos ys }
+	    return x => fn _ => x
+	    finally f => f xs
+	in f lex
```
Fix indentation.

ppolesiuk reviewed (commented):
Changes looks good, and the solution seems to work. As noticed, there are performance issues with DBL, but with upcoming changes it should work fine. I have one major remark (related to import preamble), and several minor remarks related to code formatting (see below). Moreover, the code could be better documented (this remark concerns the whole CPSPG, but a one thousand miles journey begins with a single step).

jauzatopauza committed:
Fix import issue

- Removed awkward parameter declarations which ultimately were redundant.
The user may declare the parameters themselves if they wish.
- Improved formatting.

jauzatopauza committed:
More formatting
